### PR TITLE
ISIS Powder improvements for using non-standard grouping files

### LINKED ISSUE #21259: ISIS Powder HRPD improve error message for mismatch between banks and cropping values
When applying non-standard grouping files on HRPD, it's easy to run into the error `"The number of TOF cropping values does not match the number of banks for this instrument"` which occurs, predictably, when the user supplies the wrong number of cropping values. It would be much easier to immediately see what's gone wrong if the number of banks and the number of cropping values were included in the error message.

### PR DESCRIPTION
**Pull request by josephframsay**
HRPD have a number of non-standard grouping files that they like to use. Doing this makes it reasonably easy to run into errors where there is a mismatch between the numbers of banks in the various inputs. This PR improves the readability of these errors and also enables dividing by a vanadium spline with only one bank.

**To test:**

As with all `isis_powder` PRs, if you've not set up your directories and config files, give me a poke and I'll give you a hand. Sample data is [here](https://github.com/mantidproject/mantid/files/1501831/HRPD.zip).

Create your HRPD instrument using the attached config file (you'll need to set the paths in it to suit your setup)

```
from isis_powder import HRPD
hrpd = HRPD(config_file=r"path/to/hrpd_config.yaml")
```
First, create your vanadium and do a focus normally. This should give you no errors, the output workspaces should have nice sharp peaks
```
hrpd.create_vanadium(first_cycle_run_no=68236, window="10-110")
hrpd.focus(run_number=68245, window="10-110")
```
Put `hrpd_new_072_01_corr_simm_left3_test.cal` (attached) in your calibration directory, and then try and do a focus using it:
```
hrpd.focus(run_number=68245, window="10-110",
           grouping_file_name="hrpd_new_072_01_corr_simm_left3_test.cal")
```
You should get:
```
RuntimeError: Mismatch between number of banks in vanadium and number of banks in workspace to focus
There are 3 banks for vanadium but 1 for the run
```
Now try creating your vanadium with the new grouping file and focusing with the correct number of cropping values and bin widths:
```
hrpd.create_vanadium(first_cycle_run_no=68236, window="10-110",
                     grouping_file_name="hrpd_new_072_01_corr_simm_left3_test.cal")
hrpd.focus(run_number=68245, window="10-110", focused_bin_widths=[-0.0005],
           focused_cropping_values=[(2e4, 1e5)], vanadium_normalisation=True)
```
This should not throw any errors, and you should again get nicely defined peaks. Check that the workspace `68245.raw-Results-TOF-Grp` is cropped according to the parameters in the script.

Fixes #21259 

**Release Notes** 
[Here](https://github.com/mantidproject/mantid/commit/7c5cec855bbf9d5aa71ea32e02da31a22ac64f34)

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/Diffraction/isis_powder/routines/common.py [hunk 1]
```diff
@@ -43,7 +43,9 @@ def crop_banks_using_crop_list(bank_list, crop_values_list):
 
     # Finally check the number of elements are equal
     if len(bank_list) != len(crop_values_list):
-        raise RuntimeError("The number of TOF cropping values does not match the number of banks for this instrument")
+        raise RuntimeError("The number of TOF cropping values does not match the number of banks for this instrument.\n"
+                           "{} cropping windows were supplied for {} banks".format(len(crop_values_list), len(bank_list)
+                                                                                   ))
```
Comment (samueljackson92) [inline scripts/Diffraction/isis_powder/routines/common.py]: We're calling `len` 4 times here but we could do in in 2 if we cached the length.

File: docs/source/release/v3.12.0/diffraction.rst [hunk 2]
```diff
@@ -25,6 +25,7 @@ Powder Diffraction
 - New algorithm :ref:`algm-SumOverlappingTubes` combines a detector scan for D2B into a single workspace.
 - ISIS Powder scripts for HRPD now support extra TOF windows 10-50 and 180-280
 - After calling create_vanadium and focus in ISIS Powder scripts on POLARIS, the output workspaces always contain the sample material if it is set using set_sample_material. (To view the sample material, right click the workspace and click 'Sample Material...')
+- Some usability improvements for using non-standard grouping files in ISIS_Powder
```
Comment (samueljackson92) [inline docs/source/release/v3.12.0/diffraction.rst]: What usability improvements?

File: scripts/Diffraction/isis_powder/routines/focus.py [hunk 3]
```diff
@@ -105,14 +105,30 @@ def _batched_run_focusing(instrument, perform_vanadium_norm, run_number_string,
     return output
 
 
+def _divide_one_spectrum_by_spline(spectrum, spline):
+    rebinned_spline = mantid.RebinToWorkspace(WorkspaceToRebin=spline, WorkspaceToMatch=spectrum)
+    divided = mantid.Divide(LHSWorkspace=spectrum, RHSWorkspace=rebinned_spline, OutputWorkspace=spectrum)
+    common.remove_intermediate_workspace(rebinned_spline)
```
Comment (samueljackson92) [inline scripts/Diffraction/isis_powder/routines/focus.py]: Can we use `StoreInADS=False` instead?

File: scripts/Diffraction/isis_powder/routines/focus.py [hunk 4]
```diff
@@ -105,14 +105,30 @@ def _batched_run_focusing(instrument, perform_vanadium_norm, run_number_string,
...
+    vanadium_splines = mantid.LoadNexus(Filename=spline_file_path)
+
+    if hasattr(vanadium_splines, "OutputWorkspace"):  # vanadium_splines is a group
+        vanadium_splines = vanadium_splines.OutputWorkspace
+        if len(vanadium_splines) != len(spectra_list):
+            raise RuntimeError("Mismatch between number of banks in vanadium and number of banks in workspace to focus"
+                               "\nThere are {} banks for vanadium but {} for the run".format(len(vanadium_splines),
+                                                                                             len(spectra_list)))
+        output_list = []
+        for data_ws, van_ws in zip(spectra_list, vanadium_splines):
```
Comment (samueljackson92) [inline scripts/Diffraction/isis_powder/routines/focus.py]: Perhaps:

```python
output_list = map(_divide_one_spectrum_by_spline, zip(spectra_list, vanadium_splines))
```

File: scripts/Diffraction/isis_powder/routines/focus.py [hunk 5]
```diff
@@ -105,14 +105,30 @@ def _batched_run_focusing(instrument, perform_vanadium_norm, run_number_string,
...
-        output_list.append(output_ws)
-        common.remove_intermediate_workspace(vanadium_ws)
+    vanadium_splines = mantid.LoadNexus(Filename=spline_file_path)
+
+    if hasattr(vanadium_splines, "OutputWorkspace"):  # vanadium_splines is a group
+        vanadium_splines = vanadium_splines.OutputWorkspace
+        if len(vanadium_splines) != len(spectra_list):
+            raise RuntimeError("Mismatch between number of banks in vanadium and number of banks in workspace to focus"
+                               "\nThere are {} banks for vanadium but {} for the run".format(len(vanadium_splines),
+                                                                                             len(spectra_list)))
```
Comment (samueljackson92) [inline scripts/Diffraction/isis_powder/routines/focus.py]: `len` used twice here as well.

**Review by samueljackson92 — CHANGES_REQUESTED** [review]

**Review by samueljackson92 — APPROVED** [review]

**Comment by martyngigg** [comment]
@josephframsay There are some failing system tests.

**Pull request merged**